### PR TITLE
figure03b plots of randomized data in comparison to true data

### PR DESCRIPTION
**Pull request by MrBones1102**
Added MRSA regression analysis using randomized PCA component data or shuffled statuses to see if patterns determining outcome in original data is significantly different from random.

### CODE REVIEW

**Review by aarmey — CHANGES_REQUESTED** [review]
One change, then this looks great.

**Comment by MrBones1102** [comment]
Added random orthogonal matrix. It requires a bit more computation time since it needs to generate such a large NxN matrix which I can then subset from.

File: mrsa_ca_rna/figures/figure03b.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    mrsa_data.X = scaled_X
+
+    # Explicitly convert X to np array to avoid calling shape on None
+    X = np.asarray(mrsa_data.X)
+
+    # Generate a random matrix orthogonal matrix the same size as X
+    random_SVD = ortho_group.rvs(dim=X.shape[1])
+
+    # select the first 70 components of the random SVD
+    random_pca = random_SVD[:, :70]
```
Comment (aarmey) [inline mrsa_ca_rna/figures/figure03b.py:43]: Didn't realize this only created square matrices. You can then just create a random matrix of the correct size and then use QR decomposition to orthogonalize.

File: mrsa_ca_rna/figures/figure03b.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+def genFig():
+    fig_size = (16, 4)
+    layout = {"ncols": 4, "nrows": 1}
+    ax, f, _ = setupBase(fig_size, layout)
+
+    # plot the ROC curve for random PCA
+    proba_dict, y_true = setup_figure03b()
+
+    for i, key in enumerate(proba_dict):
```
Comment (aarmey) [inline mrsa_ca_rna/figures/figure03b.py:83]: Just make a function that makes the plot for one entry of proba_dict. This would get rid of a loop inside the function above.

File: mrsa_ca_rna/figures/figure03b.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+from mrsa_ca_rna.figures.base import setupBase
+from mrsa_ca_rna.import_data import concat_datasets
+from mrsa_ca_rna.pca import perform_pca
+from mrsa_ca_rna.regression import perform_PC_LR
+
+skf = StratifiedKFold(n_splits=10)
+
+
+def setup_figure03b():
+    """Perform similar analysis of figure03a, but with random pca"""
```
Comment (aarmey) [inline mrsa_ca_rna/figures/figure03b.py:21]: Don't reference another figure in the description. Just describe what this is doing plainly.

File: mrsa_ca_rna/figures/figure03b.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    # - True MRSA data, True status
+    # - True MRSA data, Shuffled status
+    # - Random MRSA data, True status
+
+    results = {}
+
+    for data_key, data_value in mrsa_dict.items():
+        for status_key, status_value in status_dict.items():
+            key = f"{data_key}_{status_key}"
+            _, model = perform_PC_LR(data_value, status_value, return_clf=True)
```
Comment (aarmey) [inline mrsa_ca_rna/figures/figure03b.py:67]: Hasn't this function already done cross-validation?

File: mrsa_ca_rna/figures/figure03b.py (hunk 2 above)
Comment (aarmey) [reply mrsa_ca_rna/figures/figure03b.py:83]: I don't like looping to accumulate results, then looping to plot all the results. It's unnecessarily convoluted.

File: mrsa_ca_rna/figures/figure03b.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
+# main module imports
```
Comment (aarmey) [inline mrsa_ca_rna/figures/figure03b.py:1]: Make docstring for overall file.

Comment (aarmey) [inline mrsa_ca_rna/figures/figure03b.py:1]: I also think it would be good to organize your figure files a bit more. I tried to find your Pf2 analysis and had lots of trouble navigating your code. Perhaps all of this sort of analysis could have "_project_" in the name? Then the Pf2 analysis could have "_pf2_"?

**Review by aarmey — CHANGES_REQUESTED** [review]
More changes.

File: mrsa_ca_rna/figures/figure_pca_random_comparison.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    and using random data. The statuses are either shuffled or not for each case"""
+
+    # Get MRSA data
+    orig_data = concat_datasets(scale=False, tpm=True)
+    mrsa_data = orig_data[orig_data.obs["disease"] == "MRSA"].copy()
+    ca_data = orig_data[orig_data.obs["disease"] == "Candidemia"].copy()
+    y_true = mrsa_data.obs.loc[:, "status"].astype(int)
+
+    # Perform PCA on CA data to get CA components
+    _, _, ca_pca = perform_pca(ca_data.to_df())
```
Comment (aarmey) [inline mrsa_ca_rna/figures/figure_pca_random_comparison.py:51]: Why doesn't this take a number of components?

File: mrsa_ca_rna/figures/figure_pca_random_comparison.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    # scale MRSA data prior to use
+    X = mrsa_data.X
+    scaled_X = StandardScaler().fit_transform(X)
+    mrsa_data.X = scaled_X
+
+    # Explicitly convert X to np array to avoid calling shape on None
+    X = np.asarray(mrsa_data.X)
+
+    # Generate a random matrix
+    random_matrix = np.random.rand(X.shape[1], ca_pca.components_.shape[0])
```
Comment (aarmey) [inline mrsa_ca_rna/figures/figure_pca_random_comparison.py:62]: Look here: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ortho_group.html#scipy.stats.ortho_group

I like that this function is a little more explicit about what you are doing.

File: mrsa_ca_rna/figures/figure_pca_random_comparison.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    layout = {"ncols": 3, "nrows": 1}
+    ax, f, _ = setupBase(fig_size, layout)
+
+    mrsa_dict, y_shuffled, y_true = setup_figure03b()
+
+    # perform logistic regression for the following cases:
+    # - MRSA data, True status
+    # - MRSA data, Shuffled status
+    # - Random data, True status
+    true_mrsa, true_score = make_roc_curve(mrsa_dict["mrsa"], y_true, y_true)
```
Comment (aarmey) [inline mrsa_ca_rna/figures/figure_pca_random_comparison.py:91]: Why does this function take `y_true` twice?

**Review by aarmey — CHANGES_REQUESTED** [review]

File: mrsa_ca_rna/figures/figure_pca_random_comparison.py (hunk 7 above)
Comment (MrBones1102) [reply mrsa_ca_rna/figures/figure_pca_random_comparison.py:62]: We had previously used that function but it can only generate an NxN orthogonal matrix. There was excess computation time spent generating a `(X.shape[0], X.shape[0])` matrix only to then cut it down to `(X.shape[0], desired_comp)` to keep the size similar to a pca decomposition of X.

But I agree regarding the clarity of using the explicit function and I'll swap back to it.

**Review by aarmey — APPROVED** [review]
